File: src/main/java/arc/arc/hooks/lands/LandsHook.java
package arc.arc.hooks.lands;

import arc.arc.ARC;
import arc.arc.hooks.HookRegistry;
import me.angeschossen.lands.api.LandsIntegration;
import me.angeschossen.lands.api.flags.type.Flags;
import me.angeschossen.lands.api.land.Land;
import me.angeschossen.lands.api.land.LandWorld;
import me.angeschossen.lands.api.player.LandPlayer;
import org.bukkit.Chunk;
import org.bukkit.Location;
import org.bukkit.Material;
import org.bukkit.entity.Player;
import org.jetbrains.annotations.NotNull;

import java.util.UUID;
import java.util.concurrent.CompletableFuture;

public class LandsHook {

    LandsIntegration integration;

    public LandsHook() {
        integration = LandsIntegration.of(ARC.plugin);
    }

    public CompletableFuture<Location> getSpawnLocation(UUID playerUuid) {
        return integration.getOfflineLandPlayer(playerUuid)
                .thenApply(offlinePlayer -> offlinePlayer.getEditLand().getSpawn())
                .exceptionally(e -> null);
    }


    public boolean canBuild(Player player, Chunk chunk) {
        LandPlayer landPlayer = integration.getLandPlayer(player.getUniqueId());
        LandWorld landWorld = integration.getWorld(chunk.getWorld());
        if (landWorld == null) return true;
        if (landPlayer == null) return false;
        Location location = chunk.getBlock(1, 1, 1).getLocation();
        return landWorld.hasRoleFlag(landPlayer, location, Flags.BLOCK_BREAK, Material.STONE, false) &&
                landWorld.hasRoleFlag(landPlayer, location, Flags.BLOCK_PLACE, Material.STONE, false);
    }

    public boolean isClaimed(@NotNull Location location) {
        LandWorld landWorld = integration.getWorld(location.getWorld());
        if (landWorld == null) return false;
        return landWorld.getArea(location) != null;
    }
}

File: src/main/kotlin/ru/arc/hooks/lands/LandsHook.kt
package ru.arc.hooks.lands

import me.angeschossen.lands.api.LandsIntegration
import me.angeschossen.lands.api.flags.type.Flags
import org.bukkit.Chunk
import org.bukkit.Location
import org.bukkit.Material
import org.bukkit.entity.Player
import ru.arc.ARC
import java.util.UUID
import java.util.concurrent.CompletableFuture

class LandsHook {

    private val integration: LandsIntegration = LandsIntegration.of(ARC.instance)

    fun getSpawnLocation(playerUuid: UUID): CompletableFuture<Location?> =
        integration.getOfflineLandPlayer(playerUuid)
            .thenApply { it.editLand?.spawn }
            .exceptionally { null }

    fun canBuild(player: Player, chunk: Chunk): Boolean {
        val landPlayer = integration.getLandPlayer(player.uniqueId) ?: return false
        val landWorld = integration.getWorld(chunk.world) ?: return true
        val location = chunk.getBlock(1, 1, 1).location
        return landWorld.hasRoleFlag(landPlayer, location, Flags.BLOCK_BREAK, Material.STONE, false) &&
            landWorld.hasRoleFlag(landPlayer, location, Flags.BLOCK_PLACE, Material.STONE, false)
    }

    fun isClaimed(location: Location): Boolean {
        val landWorld = integration.getWorld(location.world) ?: return false
        return landWorld.getArea(location) != null
    }
}
